arangod/Replication/ReplicationClients.cpp

```diff
@@ -0,0 +1,163 @@
+    return;
+  }
+
+  if (ttl <= 0.0) {
+    ttl = replutils::BatchInfo::DefaultTimeout;
+  }
+  double const timestamp = TRI_microtime();
+  double const expires = timestamp + ttl;
+
+  WRITE_LOCKER(writeLocker, _lock);
+
+  auto it = _clients.find(clientId);
+
+  if (it == _clients.end()) {
+    // insert new client entry
+    _clients.emplace(clientId, ReplicationClientProgress(timestamp, expires, lastServedTick));
+    LOG_TOPIC("69c75", TRACE, Logger::REPLICATION)
+        << "inserting replication client entry for client '" << clientId
+        << "' using TTL " << ttl << ", last tick: " << lastServedTick;
+    return;
+  }
+
+  // update an existing client entry
+  (*it).second.lastSeenStamp = timestamp;
+  (*it).second.expireStamp = expires;
+  if (lastServedTick > 0) {
+    (*it).second.lastServedTick = lastServedTick;
+    LOG_TOPIC("47d4a", TRACE, Logger::REPLICATION)
+        << "updating replication client entry for client '" << clientId
+        << "' using TTL " << ttl << ", last tick: " << lastServedTick;
+  } else {
+    LOG_TOPIC("fce26", TRACE, Logger::REPLICATION)
+        << "updating replication client entry for client '" << clientId
+        << "' using TTL " << ttl;
+  }
+}
+
+/// @brief serialize the existing clients to a VelocyPack builder
+void ReplicationClientsProgressTracker::toVelocyPack(velocypack::Builder& builder) const {
+  READ_LOCKER(readLocker, _lock);
+
+  for (auto const& it : _clients) {
+    builder.add(VPackValue(VPackValueType::Object));
+    builder.add("serverId", VPackValue(it.first));
+
+    char buffer[21];
+    TRI_GetTimeStampReplication(it.second.lastSeenStamp, &buffer[0], sizeof(buffer));
+    builder.add("time", VPackValue(buffer));
+
+    TRI_GetTimeStampReplication(it.second.expireStamp, &buffer[0], sizeof(buffer));
+    builder.add("expires", VPackValue(buffer));
+
+    builder.add("lastServedTick", VPackValue(it.second.lastServedTick));
+    builder.close();
+  }
+}
+
+/// @brief garbage-collect the existing list of clients
+/// thresholdStamp is the timestamp before all older entries will
+/// be collected
+void ReplicationClientsProgressTracker::garbageCollect(double thresholdStamp) {
+  LOG_TOPIC("11a30", TRACE, Logger::REPLICATION)
+      << "garbage collecting replication client entries";
+
+  WRITE_LOCKER(writeLocker, _lock);
+
+  auto it = _clients.begin();
+
+  while (it != _clients.end()) {
+    if ((*it).second.expireStamp < thresholdStamp) {
+      // found an entry that is already expired
+      LOG_TOPIC("8d7db", DEBUG, Logger::REPLICATION)
+          << "removing expired replication client entry for client '" << (*it).first << "'";
+      it = _clients.erase(it);
+    } else {
+      ++it;
+    }
+  }
+}
+
+/// @brief return the lowest lastServedTick value for all clients
+/// returns UINT64_MAX in case no clients are registered
+uint64_t ReplicationClientsProgressTracker::lowestServedValue() const {
+  uint64_t value = UINT64_MAX;
```

Comment:
is uint64max a valid return value

Reply:
Yes, the comments for this API explicitly state that.

Reply:
the call site will also ignore UINT64_MAX here, as the code there is (RocksDBBackgroundThread.cpp):
```
 92           // lowestServedValue will return the lowest of the lastServedTick values stored,
 93           // or UINT64_MAX if no clients are registered
 94           minTick = std::min(minTick, vocbase.replicationClients().lowestServedValue());
```

Reply:
So returning UINT64_MAX shouldn't do any harm.